### PR TITLE
Faster dorogovtsev mendes

### PR DESCRIPTION
**Pull request by AntoineBut**
Addresses #370.

This implementation has $O(N)$ runtime, while the previous implementation had $O(N^2)$ runtime. 

The optimization comes from storing all edges added at each iteration in an indexed table, allowing to draw a random edge in $O(1)$ time, while the previous implementation had to go through (on average) half on the nodes in order to draw a random edge, which takes $O(N)$ time. 

On my computer, this implementation uses approximately 16% more memory. 

It runs faster than the previous one even on small graphs ($N = 100$).

### CODE REVIEW

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 1]
```diff
@@ -1494,24 +1494,32 @@ function dorogovtsev_mendes(
...
+
+    bag_of_edges[1] = SimpleEdge(1, 2)
+    bag_of_edges[2] = SimpleEdge(1, 3)
+    bag_of_edges[3] = SimpleEdge(2, 3)
+    index = 3
+
+    for _ in 1:(n - 3)
+        # Choose random edge from bag
+        edge = bag_of_edges[rand(rng, 1:index)]
+        u, v = edge.src, edge.dst
```
Comment (gdalle) [inline src/SimpleGraphs/generators/randgraphs.jl:1507]:
```suggestion
        u, v = src(edge), dst(edge)
```

Comment (gdalle) [reply src/SimpleGraphs/generators/randgraphs.jl:1507]: it is generally better to use accessors than to get the fields directly

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 2]
```diff
@@ -1494,24 +1494,32 @@ function dorogovtsev_mendes(
...
+
+        # Add new vertex
+        add_vertex!(g) || throw(
+            DomainError(
+                "Failed to add vertex. One possible explanation is that type $(eltype(g)) cannot represent enough vertices",
+            ),
```
Comment (gdalle) [inline src/SimpleGraphs/generators/randgraphs.jl:1513]: why is this necessary?

**Review by gdalle — CHANGES_REQUESTED** [review]
Thanks for the fix, this is nearly good to go!

File: src/SimpleGraphs/generators/randgraphs.jl (hunk 2 above)
Comment (AntoineBut) [reply src/SimpleGraphs/generators/randgraphs.jl:1513]: After adding this I realized that it is currently note possible to use this function to generate a graph where the node types are different than Int64. Making this possible would better fit in a dedicated issue I think (and of coherence this would need to be done with other similar generators).

Currently, only this function allows for user-defined node types : 

`function SimpleDiGraph{T}(
    nv::Integer,
    ne::Integer;
    rng::Union{Nothing,AbstractRNG}=nothing,
    seed::Union{Nothing,Integer}=nothing,
) where {T<:Integer}`

If in the future we allow this kind of freedom, this checks makes sense as it could happen that someone tries to generate a graph too large for its nodes to be represented with the given type (this probably won't happen with Int64 though).

So as of now it isn't necessary.

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 3]
```diff
@@ -1494,24 +1494,32 @@ function dorogovtsev_mendes(
...
+        # Choose random edge from bag
+        edge = bag_of_edges[rand(rng, 1:index)]
+        u, v = src(edge), dst(edge)
+
+        # Add new vertex
+        add_vertex!(g) || throw(
+            DomainError(
+                "Failed to add vertex. One possible explanation is that type $(eltype(g)) cannot represent enough vertices",
+            ),
+        )
```
Comment (gdalle) [inline src/SimpleGraphs/generators/randgraphs.jl:1514]:
```suggestion
        add_vertex!(g)
```

**Review by gdalle — CHANGES_REQUESTED** [review]

**Review by gdalle — APPROVED** [review]
